File: src/frontend/nfse/lib/app/data/model/os_abertura_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfse/app/infra/infra_imports.dart';
import 'package:nfse/app/data/model/model_imports.dart';
import 'package:intl/intl.dart';

class OsAberturaModel {
	int? id;
	int? idOsStatus;
	int? idColaborador;
	int? idCliente;
	String? numero;
	DateTime? dataInicio;
	String? horaInicio;
	DateTime? dataPrevisao;
	String? horaPrevisao;
	DateTime? dataFim;
	String? horaFim;
	String? nomeContato;
	String? foneContato;
	String? observacaoCliente;
	String? observacaoAbertura;
	ViewPessoaColaboradorModel? viewPessoaColaboradorModel;
	ViewPessoaClienteModel? viewPessoaClienteModel;

	OsAberturaModel({
		this.id,
		this.idOsStatus,
		this.idColaborador,
		this.idCliente,
		this.numero,
		this.dataInicio,
		this.horaInicio,
		this.dataPrevisao,
		this.horaPrevisao,
		this.dataFim,
		this.horaFim,
		this.nomeContato,
		this.foneContato,
		this.observacaoCliente,
		this.observacaoAbertura,
		this.viewPessoaColaboradorModel,
		this.viewPessoaClienteModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'numero',
		'data_inicio',
		'hora_inicio',
		'data_previsao',
		'hora_previsao',
		'data_fim',
		'hora_fim',
		'nome_contato',
		'fone_contato',
		'observacao_cliente',
		'observacao_abertura',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Numero',
		'Data Inicio',
		'Hora Inicio',
		'Data Previsao',
		'Hora Previsao',
		'Data Fim',
		'Hora Fim',
		'Nome Contato',
		'Fone Contato',
		'Observacao Cliente',
		'Observacao Abertura',
	];

	OsAberturaModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idOsStatus = jsonData['idOsStatus'];
		idColaborador = jsonData['idColaborador'];
		idCliente = jsonData['idCliente'];
		numero = jsonData['numero'];
		dataInicio = jsonData['dataInicio'] != null ? DateTime.tryParse(jsonData['dataInicio']) : null;
		horaInicio = jsonData['horaInicio'];
		dataPrevisao = jsonData['dataPrevisao'] != null ? DateTime.tryParse(jsonData['dataPrevisao']) : null;
		horaPrevisao = jsonData['horaPrevisao'];
		dataFim = jsonData['dataFim'] != null ? DateTime.tryParse(jsonData['dataFim']) : null;
		horaFim = jsonData['horaFim'];
		nomeContato = jsonData['nomeContato'];
		foneContato = jsonData['foneContato'];
		observacaoCliente = jsonData['observacaoCliente'];
		observacaoAbertura = jsonData['observacaoAbertura'];
		viewPessoaColaboradorModel = jsonData['viewPessoaColaboradorModel'] == null ? ViewPessoaColaboradorModel() : ViewPessoaColaboradorModel.fromJson(jsonData['viewPessoaColaboradorModel']);
		viewPessoaClienteModel = jsonData['viewPessoaClienteModel'] == null ? ViewPessoaClienteModel() : ViewPessoaClienteModel.fromJson(jsonData['viewPessoaClienteModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idOsStatus'] = idOsStatus != 0 ? idOsStatus : null;
		jsonData['idColaborador'] = idColaborador != 0 ? idColaborador : null;
		jsonData['idCliente'] = idCliente != 0 ? idCliente : null;
		jsonData['numero'] = numero;
		jsonData['dataInicio'] = dataInicio != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataInicio!) : null;
		jsonData['horaInicio'] = horaInicio;
		jsonData['dataPrevisao'] = dataPrevisao != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataPrevisao!) : null;
		jsonData['horaPrevisao'] = horaPrevisao;
		jsonData['dataFim'] = dataFim != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataFim!) : null;
		jsonData['horaFim'] = horaFim;
		jsonData['nomeContato'] = nomeContato;
		jsonData['foneContato'] = foneContato;
		jsonData['observacaoCliente'] = observacaoCliente;
		jsonData['observacaoAbertura'] = observacaoAbertura;
		jsonData['viewPessoaColaboradorModel'] = viewPessoaColaboradorModel?.toJson;
		jsonData['viewPessoaClienteModel'] = viewPessoaClienteModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idOsStatus = plutoRow.cells['idOsStatus']?.value;
		idColaborador = plutoRow.cells['idColaborador']?.value;
		idCliente = plutoRow.cells['idCliente']?.value;
		numero = plutoRow.cells['numero']?.value;
		dataInicio = Util.stringToDate(plutoRow.cells['dataInicio']?.value);
		horaInicio = plutoRow.cells['horaInicio']?.value;
		dataPrevisao = Util.stringToDate(plutoRow.cells['dataPrevisao']?.value);
		horaPrevisao = plutoRow.cells['horaPrevisao']?.value;
		dataFim = Util.stringToDate(plutoRow.cells['dataFim']?.value);
		horaFim = plutoRow.cells['horaFim']?.value;
		nomeContato = plutoRow.cells['nomeContato']?.value;
		foneContato = plutoRow.cells['foneContato']?.value;
		observacaoCliente = plutoRow.cells['observacaoCliente']?.value;
		observacaoAbertura = plutoRow.cells['observacaoAbertura']?.value;
		viewPessoaColaboradorModel = ViewPessoaColaboradorModel();
		viewPessoaColaboradorModel?.nome = plutoRow.cells['viewPessoaColaboradorModel']?.value;
		viewPessoaClienteModel = ViewPessoaClienteModel();
		viewPessoaClienteModel?.nome = plutoRow.cells['viewPessoaClienteModel']?.value;
	}	

	OsAberturaModel clone() {
		return OsAberturaModel(
			id: id,
			idOsStatus: idOsStatus,
			idColaborador: idColaborador,
			idCliente: idCliente,
			numero: numero,
			dataInicio: dataInicio,
			horaInicio: horaInicio,
			dataPrevisao: dataPrevisao,
			horaPrevisao: horaPrevisao,
			dataFim: dataFim,
			horaFim: horaFim,
			nomeContato: nomeContato,
			foneContato: foneContato,
			observacaoCliente: observacaoCliente,
			observacaoAbertura: observacaoAbertura,
		);			
	}

	
}

File: backend/nfse/src/model/os_abertura_model.py
from src import db
from src.model.view_pessoa_colaborador_model import ViewPessoaColaboradorModel
from src.model.view_pessoa_cliente_model import ViewPessoaClienteModel


class OsAberturaModel(db.Model):
    __tablename__ = 'os_abertura'

    id = db.Column(db.Integer, primary_key=True)
    numero = db.Column(db.String(20))
    data_inicio = db.Column(db.DateTime)
    hora_inicio = db.Column(db.String(8))
    data_previsao = db.Column(db.DateTime)
    hora_previsao = db.Column(db.String(8))
    data_fim = db.Column(db.DateTime)
    hora_fim = db.Column(db.String(8))
    nome_contato = db.Column(db.String(50))
    fone_contato = db.Column(db.String(15))
    observacao_cliente = db.Column(db.Text)
    observacao_abertura = db.Column(db.Text)
    id_os_status = db.Column(db.Integer, db.ForeignKey('os_status.id'))
    id_colaborador = db.Column(db.Integer, db.ForeignKey('view_pessoa_colaborador.id'))
    id_cliente = db.Column(db.Integer, db.ForeignKey('view_pessoa_cliente.id'))

    view_pessoa_colaborador_model = db.relationship('ViewPessoaColaboradorModel', foreign_keys=[id_colaborador])
    view_pessoa_cliente_model = db.relationship('ViewPessoaClienteModel', foreign_keys=[id_cliente])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_os_status = data.get('idOsStatus')
        self.id_colaborador = data.get('idColaborador')
        self.id_cliente = data.get('idCliente')
        self.numero = data.get('numero')
        self.data_inicio = data.get('dataInicio')
        self.hora_inicio = data.get('horaInicio')
        self.data_previsao = data.get('dataPrevisao')
        self.hora_previsao = data.get('horaPrevisao')
        self.data_fim = data.get('dataFim')
        self.hora_fim = data.get('horaFim')
        self.nome_contato = data.get('nomeContato')
        self.fone_contato = data.get('foneContato')
        self.observacao_cliente = data.get('observacaoCliente')
        self.observacao_abertura = data.get('observacaoAbertura')

    def serialize(self):
        return {
            'id': self.id,
            'idOsStatus': self.id_os_status,
            'idColaborador': self.id_colaborador,
            'idCliente': self.id_cliente,
            'numero': self.numero,
            'dataInicio': self.data_inicio.isoformat(),
            'horaInicio': self.hora_inicio,
            'dataPrevisao': self.data_previsao.isoformat(),
            'horaPrevisao': self.hora_previsao,
            'dataFim': self.data_fim.isoformat(),
            'horaFim': self.hora_fim,
            'nomeContato': self.nome_contato,
            'foneContato': self.fone_contato,
            'observacaoCliente': self.observacao_cliente,
            'observacaoAbertura': self.observacao_abertura,
            'viewPessoaColaboradorModel': self.view_pessoa_colaborador_model.serialize() if self.view_pessoa_colaborador_model else None,
            'viewPessoaClienteModel': self.view_pessoa_cliente_model.serialize() if self.view_pessoa_cliente_model else None,
        }